src/main/java/org/opensearch/security/auth/PrincipalAttributeExtension.java

```diff
@@ -0,0 +1,22 @@
+package org.opensearch.security.auth;
+
+import org.opensearch.plugin.wlm.spi.AttributeExtension;
+import org.opensearch.rule.attribute_extractor.AttributeExtractor;
+import org.opensearch.security.OpenSearchSecurityPlugin;
+
+/**
+ * Extension that extracts the principal from the thread context
+ */
+public class PrincipalAttributeExtension implements AttributeExtension {
+
+    private final OpenSearchSecurityPlugin plugin;
+
+    public PrincipalAttributeExtension(OpenSearchSecurityPlugin plugin) {
```

Comment:
Note to self: the relevant code in core for this is here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/plugins/PluginsService.java#L686-L697